[flang][docs] Update runtime trampoline documentation (#215233)
llvm/llvm-project: flang/docs/FAQ.md

### Why do I get a warning or an error about an executable stack?

This occurs because Flang's default implementation of pointers to internal procedures requires an executable stack.

An internal procedure has a "host scope", which is the scope in which it is contained.
It can access variables defined in that host scope.
When an internal procedure is referenced from outside its host scope (for example, via a procedure pointer), the implementation must ensure that it can still access variables from the host scope.

> To achieve this, the default implementation of Flang generates a small piece of code, called a "trampoline", on the stack.

Reviewer: ```suggestion
By default, Flang achieves this by generating a small piece of code, called a "trampoline", on the stack.
```

When the procedure is called, this trampoline is executed.

> The trampoline is on the stack, so the stack itself must be executable.

Reviewer: ```suggestion
Since the trampoline is on the stack, the stack itself must be executable.
```

```

Since LLVM 23, Flang has a new compiler flag `-fsafe-trampoline`, which is disabled by default and is currently supported on x86-64 and AArch64 targets.

> If this is enabled, Flang uses a runtime-managed trampoline pool outside the stack, with executable code stored separately from writeable per-trampoline data.

Reviewer: Rather than saying "outside the stack", make it clear that this does not require an executable stack.

```suggestion
If this is enabled, Flang uses a runtime-managed trampoline pool, with executable code stored separately from writeable per-trampoline data. This does not require an executable stack.
```
